Add clear sequential word command (#415)
FuelLabs/fuel-specs: specs/vm/instruction_set.md

`messageID` is added to the `OutputMessage` Merkle tree as part of block header.
TODO: document output messages merkle tree construction and maintenance and link here

### SCWQ: State clear sequential 32 byte slots

|             |                                                                               |
|-------------|-------------------------------------------------------------------------------|
| Description | A sequential series of 32 bytes is cleared from the current contract's state. |

> | Operation   | ```STATE[MEM[$rA, 32 * rB]] = None;```                                        |

Reviewer: I'm assuming the intention is to provide a starting key and delete the next `n` entries, rather than providing an array of keys to bulk delete from the db?

```suggestion
| Operation   | ```STATE[MEM[$rA, 32], 32  * $rB] = None;```                                        |
```

| Syntax      | `scwq $rA, $rB`                                                               |
| Encoding    | `0x00 rA rB - -`                                                              |
| Notes       |                                                                               |

Panic if:

> - `$rA + 32 * rB` overflows

Reviewer: [nit] inconsistent formatting:`$rA` has `$` but `rB` does not

> - `$rA + 32 * $rB` overflows

Reviewer 2: Wait, you don't need to multiple by `$rB` here. `MEM[$rA, 32]` holds the 32 byte slot index and that's it.

Reviewer: I made the same comment here: https://github.com/FuelLabs/fuel-specs/pull/415#discussion_r993929556

Author: fixed

> - `$rA + 32 * rB > VM_MAX_RAM`

Reviewer: Since we are clearing data from the database (state) instead of VM memory, it seems like this check isn't applicable.

Author: Those checks exist for SWW and SWWQ. Should those also not exist? Or are you saying that they are getting their input values from VM Memory so the check is necessary there but our input value is already known?

Reviewer: Yeah the difference is that SWW and SWWQ are copying bytes from db into memory, and we need to ensure the memory range withing the VM is valid. 

Clearing contiguous state is different, because we are specifying data ranges inside of the DB instead of VM memory. I edited a previous comment to help clarify the intent of how I think this would be used but it might've been resolved before you saw it:

> The reason for this being that when data in sway is stored to the db, it is packed into contiguous 32 byte storage slots with an ascendingly increasing storage key. So if you wanted to remove an entire struct from the db in one operation, you could just provide the starting address and the number of slots that were used to store it.

https://github.com/FuelLabs/fuel-specs/pull/415#discussion_r993932158

Reviewer: Also realized, you meant SRW and SRWQ right? For SWW and SWWQ, it doesn't seems like they have any `32 * d` type ops yet.

> - `$rA + 32 * $rB > VM_MAX_RAM`

Reviewer: You resolved my previous comment about this, but I don't see any replies or updates to the PR which resolve this?

- `$fp == 0` (in the script context)

> 

Reviewer: Should we also set `rB` here to indicate whether the initial key actually existed?

Reviewer 2: Unlike x86, we don't re-use registers anywhere in the instruction set. So if we want to return something, we should use a third register.

Reviewer: yeah I was assuming we'd shift the current use of rB into rC to be more consistent with the other recent changes

Author: fixed

`messageID` is added to the `OutputMessage` Merkle tree as part of block header.
TODO: document output messages merkle tree construction and maintenance and link here

### SCWQ: State clear sequential 32 byte slots

|             |                                                                               |
|-------------|-------------------------------------------------------------------------------|
| Description | A sequential series of 32 bytes is cleared from the current contract's state. |
| Operation   | ```STATE[MEM[$rA, 32], 32 * $rC] = None;```                                   |
| Syntax      | `scwq $rA, $rB, $rC`                                                          |
| Encoding    | `0x00 rA rB rC -`                                                             |
| Notes       |                                                                               |

Panic if:

- `$rA + 32` overflows
- `$rA + 32 > VM_MAX_RAM`
- `$rB` is a [reserved register](./main.md#semantics)
- `$fp == 0` (in the script context)

> Register `rB` will be populated with `false` if the first word was already unset (default) and `true` if the value is set.

Reviewer 2: ```suggestion
Register `rB` will be set to `false` if the first storage slot was already unset (default) and `true` if the slot was set.
```
